src/verifast1.ml

```diff
@@ -3189,18 +3255,72 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
         let ms = get_methods tn g in
         if ms = [] then on_fail () else
         let argtps = List.map (fun e -> let (_, tp, _) = (check e) in tp) args in
+        let ms = List.map (fun (sign, (tn', lm, gh, rt, xmap, pre, post, epost, terminates, fb', v, abstract, tparams)) ->
+            (* Build a list of type arguments, either from the ones that got passed, or the ones you have to assume *)
+            let targenv = 
+              (if gh == Real && (List.length real_type_arguments) > 0 
+                then begin
+                let type_arguments_count = List.length real_type_arguments in
+                let tparams_count = List.length tparams in
+                if type_arguments_count <> tparams_count
+                  then static_error l 
+                    (Printf.sprintf "Wrong amount of type arguments provided for method: %s. Expected: %s Actual: %s" 
+                      g
+                      (string_of_int tparams_count) 
+                      (string_of_int type_arguments_count)
+                     ) None
+                  else List.map2 (fun a b -> (a,b)) tparams real_type_arguments 
+                end
+                else (*Build a list of type arguments, inferred from the signature *)
+                (* Only bother mapping if the amount of arguments given fits the sign, otherwise the method won't be a match anyway*)
+                if (List.length argtps) <> (List.length sign) 
+                  then []
+                  else 
+                  let mappedSign = 
+                    let rec findMapping t t0 = match (t,t0) with 
+                      | (RealTypeParam t,_) -> [Some((t,t0))]
+                      | (ArrayType t, ArrayType t0) -> findMapping t t0
+                      | (ObjType (n,ts), ObjType (n0,ts0)) -> if (List.length ts) <> (List.length ts) 
+                        then static_error l (Printf.sprintf "Passed type is incompatible. Expected %s. Real %s."
+                          (string_of_type (ObjType (n,ts)))
+                          (string_of_type (ObjType (n0, ts0)))) None
+                        else List.concat (List.map2 findMapping ts ts0)
+                      | (_,_) -> [None] in
+                        List.concat (List.map2 findMapping sign argtps) in
+                  let filtered_mapped_sign = List.filter (fun entry -> match entry with | Some(s) -> true | None -> false) mappedSign in
+                    List.map (fun entry -> match entry with Some(s) -> s) filtered_mapped_sign)
+                    (* TODO: we are making assumptions, check the common parent if one type parameter maps on multiple argument types, otherwise something is wrong*)
+            in
+            let rec extract_targs targenv tparams = 
+              match targenv with
+                (p,a)::t -> if (List.mem p tparams) then extract_targs t tparams else a::(extract_targs t (p::tparams))
+                | [] -> [] in

```

Comment:
The resulting list of type arguments may be in the wrong order, and it may not be long enough?